### PR TITLE
Fix metric-server RBAC manifests

### LINKED ISSUE #72324: Failed to deploy metrics-server on k8s v1.13 with default resource-reader.yaml
**What happened**: container metrics-server failed. error message in container log:
request failed - "403 Forbidden", response: "Forbidden (user=system:serviceaccount:kube-system:metrics-server, verb=get, resource=nodes, subresource=stats)"

**What you expected to happen**: should succeed.

**How to reproduce it (as minimally and precisely as possible)**: kubectl apply -f 6 yaml files in https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/metrics-server
(should edit metrics-server-deployment.yaml to fit environment first)

**Anything else we need to know?**:
to fix, just add nodes/stats resource in default resource-reader.yaml:
```
  resources:
  - pods
  - nodes
  - nodes/stats
  - namespaces
```

**Environment**: 
- Kubernetes version (use `kubectl version`): v1.13.0
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

/kind bug

### PR DESCRIPTION
**Pull request by mvladev**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

`metrics-server` needs additional permissions and added aggregated read cluster role

**Which issue(s) this PR fixes**:

Fixes #72324

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

/assign @DirectXMan12

### CODE REVIEW

File: cluster/addons/metrics-server/aggregated-metrics-reader.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: system:aggregated-metrics-reader
+  labels:
+    kubernetes.io/cluster-service: "true"
+    addonmanager.kubernetes.io/mode: Reconcile
+    rbac.authorization.k8s.io/aggregate-to-view: "true"
+    rbac.authorization.k8s.io/aggregate-to-edit: "true"
+    rbac.authorization.k8s.io/aggregate-to-admin: "true"
```
Comment (liggitt) [inline cluster/addons/metrics-server/aggregated-metrics-reader.yaml]: does the metrics server allow namespaced access to pod metrics? if not, these permissions should not be added to the admin/edit/view roles, which are intended to be used by namespaced users

File: cluster/addons/metrics-server/resource-reader.yaml [hunk 2]
```diff
@@ -23,7 +24,6 @@ rules:
   verbs:
   - get
   - list
-  - update
```
Comment (liggitt) [inline cluster/addons/metrics-server/resource-reader.yaml]: this was added in https://github.com/kubernetes/kubernetes/commit/17c9e344599105b77f38ca7e54aff5dd5e457603#diff-e56c72ed8f82cf9883ae09e465a4675d and while I agree this is overly broad, we need to make sure that the use case that required it is still accomodated

Comment (liggitt) [reply cluster/addons/metrics-server/resource-reader.yaml]: cc @piosz

**Comment by liggitt** [comment]
/hold

for question about namespaced metrics, and for question about the pod-nanny use case for updating deployments

File: cluster/addons/metrics-server/resource-reader.yaml (hunk 2 above)
Comment (mvladev) [reply cluster/addons/metrics-server/resource-reader.yaml]: I can probably change it to only allow updates to that specific deployment.

File: cluster/addons/metrics-server/aggregated-metrics-reader.yaml (hunk 1 above)
Comment (mvladev) [reply cluster/addons/metrics-server/aggregated-metrics-reader.yaml]: Those metrics are namespaced:

```json
/apis/metrics.k8s.io/v1beta1/namespaces/kube-system/pods

{

    "kind": "PodMetricsList",
    "apiVersion": "metrics.k8s.io/v1beta1",
    "metadata": {
        "selfLink": "/apis/metrics.k8s.io/v1beta1/namespaces/kube-system/pods"
    },
    "items": [
        {
            "metadata": {
                "name": "kube-proxy-tgq75",
                "namespace": "kube-system",
                "selfLink": "/apis/metrics.k8s.io/v1beta1/namespaces/kube-system/pods/kube-proxy-tgq75",
                "creationTimestamp": "2018-12-26T22:13:45Z"
            },
            "timestamp": "2018-12-26T22:12:47Z",
            "window": "30s",
            "containers": [
                {
                    "name": "kube-proxy",
                    "usage": {
                        "cpu": "2584225n",
                        "memory": "27296Ki"
                    }
                }
            ]
        },
        {
            "metadata": {
                "name": "privileged-pod",
                "namespace": "kube-system",
                "selfLink": "/apis/metrics.k8s.io/v1beta1/namespaces/kube-system/pods/privileged-pod",
                "creationTimestamp": "2018-12-26T22:13:45Z"
            },
            "timestamp": "2018-12-26T22:12:49Z",
            "window": "30s",
            "containers": [
                {
                    "name": "busybox",
                    "usage": {
                        "cpu": "0",
                        "memory": "2236Ki"
                    }
                }
            ]
        }
]}

File: cluster/addons/metrics-server/resource-reader.yaml (hunk 2 above)
Comment (mvladev) [reply cluster/addons/metrics-server/resource-reader.yaml]: Yes, you are right.

The nanny side-car needs as permission is to `list` and `watch` for `Nodes`, `get` on `Pod` (we can't know the name in advance), `get` and `update` on `Deployment` (for which we can use the deployment name and `Role`). 
https://github.com/kubernetes/autoscaler/blob/cd56380d76de1587c6c71c6cb62c5dfb3bd1f0e3/addon-resizer/nanny/kubernetes_client.go#L32-L111

I'm not sure how the nanny sidecar is working currently as it it clearly requires `get` to `Pods` [in the namespace](https://github.com/kubernetes/autoscaler/blob/cd56380d76de1587c6c71c6cb62c5dfb3bd1f0e3/addon-resizer/nanny/kubernetes_client.go#L56), but I don't see any bindings giving it access to it.

**Comment by mvladev** [comment]
@liggitt reduced granted permissions

- removed `namespaces` - it's not used by `metrics-server` or `nanny`
- introduce a new `Role` for `get` and `update` the `metrics-server` deployment (used by `nanny`)

File: cluster/addons/metrics-server/aggregated-metrics-reader.yaml (hunk 1 above)
Comment (mvladev) [reply cluster/addons/metrics-server/aggregated-metrics-reader.yaml]: Also https://github.com/kubernetes-incubator/metrics-server/blob/ae42a7df60ec08f1fe653ed99f504e85cd34d5cd/pkg/storage/podmetrics/reststorage.go#L162-L164

**Comment by mvladev** [comment]
> for question about namespaced metrics

Those metrics are namespaces - see [storage.go](https://github.com/kubernetes-incubator/metrics-server/blob/92d8412788e27ee669d38f21f20bad5342211884/pkg/storage/podmetrics/reststorage.go#L166-L168).

> and for question about the pod-nanny use case for updating deployments 

I think that those permissions should be sufficient. Ping @piosz and @wojtek-t for clarification.

**Comment by liggitt** [comment]
/hold cancel
/approve

will leave lgtm to @DirectXMan12

**Pull request closed without merging**